flashcards/backend/users/tests/test_views.py

```diff
@@ -0,0 +1,165 @@
+import pytest
+from rest_framework import status
+from rest_framework.test import APIRequestFactory
+from rest_framework_simplejwt.views import TokenObtainPairView, TokenRefreshView, TokenVerifyView
+
+from users.views import CustomUserCreateView, CustomUserLogoutViewWithBlacklistTokenReset, CustomUserUpdatePasswordView
+
+
+def test_create_user_by_api(db, api_rf):
+    url = "/api/v1/users/create/"
+    view = CustomUserCreateView.as_view()
+    new_user = {"email": "d@d.pl", "password": "abcdesfg"}
+
+    request = api_rf.post(url, new_user, format="json")
+    request2 = api_rf.post(url, new_user, format="json")
+
+    response = view(request)
+
+    assert response.status_code == status.HTTP_201_CREATED
+    assert response.data.get("email") == "d@d.pl"
+    assert len(response.data.get("id")) != ""
+
+
+def test_create_user_by_api_duplicate_error(db, api_rf):
+    url = "/api/v1/users/create/"
+    view = CustomUserCreateView.as_view()
+    new_user = {"email": "d@d.pl", "password": "abcdesfg"}
+
+    request = api_rf.post(url, new_user, format="json")
+    request2 = api_rf.post(url, new_user, format="json")
+
+    response = view(request)
+    response2 = view(request2)
+
+    assert response2.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_create_user_by_api_invalid_email(db, api_rf):
+    url = "/api/v1/users/create/"
+    view = CustomUserCreateView.as_view()
+    new_user = {"email": "d.pl", "password": "abcdesfg"}
+
+    request = api_rf.post(url, new_user, format="json")
+    response = view(request)
+
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_create_user_by_api_invalid_password(db, api_rf):
+    url = "/api/v1/users/create/"
+    view = CustomUserCreateView.as_view()
+    new_user = {"email": "test@test.pl"}
+    request = api_rf.post(url, new_user, format="json")
+    response = view(request)
+    assert response.status_code == status.HTTP_400_BAD_REQUEST
+
+
+def test_login_user_by_api_invalid_credentials(db, api_rf):
+    url = "/api/v1/users/token/"
+    view = TokenObtainPairView.as_view()
+    new_user = {"email": "test@test.pl", "password": "abcdesfg"}
+    request = api_rf.post(url, new_user, format="json")
+    response = view(request)
+
+    assert response.status_code == status.HTTP_401_UNAUTHORIZED
+    assert response.data.get("detail") == "No active account found with the given credentials"
+
+
+@pytest.mark.skip(reason="User created inactive.")
+def test_login_user_by_api_create_and_token_endpoint_success(db, api_rf, new_user_admin):
```

Comment:
There is a lot that happens in this test, I think it is a bit too much. When one assertion fails we won't know which one. and I don't think you should handle registration here. Your story is not saying anything about registration, login and all this 😅 [Link to your story](https://azeno.youtrack.cloud/issue/AF-32/Dashboard-View)